hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala

```diff
@@ -244,6 +245,65 @@ object SNDArray {
     }
   }
 
+  def copy(cb: EmitCodeBuilder, _X: SNDArrayCode, _Y: SNDArrayCode): Unit = {
+    val X = _X.memoize(cb, "copy_X")
+    val Y = _Y.memoize(cb, "copy_Y")
+    assertVector(X, Y)
+    val n = X.shapes(cb)(0)
+    cb.ifx(Y.shapes(cb)(0).cne(n), cb._fatal("copy: vectors have different sizes"))
+
+    val ldX = (X.strides(cb)(0).toI >> 3).max(1)
+    val ldY = (Y.strides(cb)(0).toI >> 3).max(1)
+    cb += Code.invokeScalaObject5[Int, Long, Int, Long, Int, Unit](BLAS.getClass, "dcopy",
+      n.toI,
+      X.firstDataAddress(cb), ldX,
+      Y.firstDataAddress(cb), ldY)
+  }
+
+  def scale(cb: EmitCodeBuilder, alpha: SCode, X: SNDArrayCode): Unit =
+    scale(cb, alpha.asInstanceOf[SFloat64Code].code, X)
+
+  def scale(cb: EmitCodeBuilder, alpha: Code[Double], _X: SNDArrayCode): Unit = {
+    val X = _X.memoize(cb, "copy_X")
+    assertVector(X)
+    val n = X.shapes(cb)(0)
+
+    val ldX = (X.strides(cb)(0).toI >> 3).max(1)
+    cb += Code.invokeScalaObject4[Int, Double, Long, Int, Unit](BLAS.getClass, "dscal",
+      n.toI, alpha, X.firstDataAddress(cb), ldX)
+  }
+
+  def gemv(cb: EmitCodeBuilder, trans: String, A: SNDArrayCode, X: SNDArrayCode, Y: SNDArrayCode): Unit = {
+    gemv(cb, trans, 1.0, A, X, 1.0, Y)
+  }
+
+  def gemv(cb: EmitCodeBuilder, trans: String, alpha: Code[Double], _A: SNDArrayCode, _X: SNDArrayCode, beta: Code[Double], _Y: SNDArrayCode): Unit = {
+    val A = _A.memoize(cb, "copy_A")
+    val X = _X.memoize(cb, "copy_X")
+    val Y = _Y.memoize(cb, "copy_Y")
+
+    assertMatrix(A)
+    assertColMajor(cb, A)
+    assertVector(X, Y)
+
+    val Seq(m, n) = A.shapes(cb)
+    if (trans == "N")
+      cb.ifx(X.shapes(cb)(0).cne(n) || Y.shapes(cb)(0).cne(m), cb._fatal("gemv: incompatible dimensions"))
+    else
+      cb.ifx(X.shapes(cb)(0).cne(m) || Y.shapes(cb)(0).cne(n), cb._fatal("gemv: incompatible dimensions"))
+
+    val ldA = (A.strides(cb)(1).toI >> 3).max(1)
+    val ldX = (X.strides(cb)(0).toI >> 3).max(1)
+    val ldY = (Y.strides(cb)(0).toI >> 3).max(1)
```

Comment:
Don't have to do anything with this comment, but I wonder if we shouldn't just have a `byteStride(i)` function on `SNDArrayValue` that looks up the stride at index `i` and does the bitshift?

Reply:
Yeah, I like that. In fact, I'd like it if the SNDArray interface completely abstracted over whether strides are stored as bytes or elements.